File: server/resources/js/game/rules.ts
/**
 * The admin-tunable game clock (App\Support\GameRules): day and night lengths
 * and the zombie schedule. The web game gets them as a page prop, the native
 * client from /api/rules, and a host passes its own to joiners in the welcome
 * so a match runs on one set of rules whatever the admin changes meanwhile.
 */
export interface GameRules {
  daySeconds: number
  nightSeconds: number
  /** zombies on night 1 */
  zombiesFirstNight: number
  /** added every night after the first */
  zombiesPerNight: number
  /** seconds after sunset before the first group appears */
  spawnDelaySeconds: number
  /** the groups are spread over this share of the night, 0..1 */
  spawnWindow: number
}

export const DEFAULT_RULES: Readonly<GameRules> = Object.freeze({
  daySeconds: 900,
  nightSeconds: 300,
  zombiesFirstNight: 8,
  zombiesPerNight: 6,
  spawnDelaySeconds: 2,
  spawnWindow: 0.7,
})

/** what the server sends (percent instead of a fraction) */
export interface GameRulesWire {
  daySeconds: number
  nightSeconds: number
  zombiesFirstNight: number
  zombiesPerNight: number
  spawnDelaySeconds: number
  spawnWindowPercent: number
}

const num = (v: unknown, fallback: number, min: number, max: number): number =>
  typeof v === 'number' && Number.isFinite(v) ? Math.min(max, Math.max(min, v)) : fallback

/** Rules from the server (or a welcome), with anything missing or absurd replaced by the default. */
export function parseRules(raw: Partial<GameRulesWire> | null | undefined): GameRules {
  const r = raw ?? {}
  return {
    daySeconds: num(r.daySeconds, DEFAULT_RULES.daySeconds, 30, 7200),
    nightSeconds: num(r.nightSeconds, DEFAULT_RULES.nightSeconds, 30, 7200),
    zombiesFirstNight: num(r.zombiesFirstNight, DEFAULT_RULES.zombiesFirstNight, 0, 200),
    zombiesPerNight: num(r.zombiesPerNight, DEFAULT_RULES.zombiesPerNight, 0, 100),
    spawnDelaySeconds: num(r.spawnDelaySeconds, DEFAULT_RULES.spawnDelaySeconds, 0, 600),
    spawnWindow: num(r.spawnWindowPercent, DEFAULT_RULES.spawnWindow * 100, 5, 100) / 100,
  }
}

export function rulesToWire(rules: GameRules): GameRulesWire {
  return {
    daySeconds: rules.daySeconds,
    nightSeconds: rules.nightSeconds,
    zombiesFirstNight: rules.zombiesFirstNight,
    zombiesPerNight: rules.zombiesPerNight,
    spawnDelaySeconds: rules.spawnDelaySeconds,
    spawnWindowPercent: Math.round(rules.spawnWindow * 100),
  }
}

/** the rules the page loaded with; the host's welcome overrides them for a joiner */
let current: GameRules = DEFAULT_RULES

export const setCurrentRules = (rules: GameRules): void => { current = rules }
export const currentRules = (): GameRules => current

File: client/lib/game/rules.dart
/// The admin-tunable game clock — twin of `game/rules.ts` and
/// `App\Support\GameRules`: day and night lengths and the zombie schedule.
/// Loaded from `GET /api/rules` before a match; a joiner takes the host's
/// from the welcome so one match runs on one set of rules.
library;

final class GameRules {
  const GameRules({
    this.daySeconds = 900,
    this.nightSeconds = 300,
    this.zombiesFirstNight = 8,
    this.zombiesPerNight = 6,
    this.spawnDelaySeconds = 2,
    this.spawnWindow = 0.7,
  });

  /// what the server sends (percent instead of a fraction); anything missing
  /// or absurd becomes the default
  factory GameRules.fromJson(Map<String, dynamic>? j) {
    const d = GameRules();
    double pick(String key, double fallback, double min, double max) {
      final v = j?[key];
      if (v is! num) return fallback;
      return v.toDouble().clamp(min, max);
    }

    return GameRules(
      daySeconds: pick('daySeconds', d.daySeconds, 30, 7200),
      nightSeconds: pick('nightSeconds', d.nightSeconds, 30, 7200),
      zombiesFirstNight: pick('zombiesFirstNight', 8, 0, 200).round(),
      zombiesPerNight: pick('zombiesPerNight', 6, 0, 100).round(),
      spawnDelaySeconds: pick('spawnDelaySeconds', d.spawnDelaySeconds, 0, 600),
      spawnWindow:
          pick('spawnWindowPercent', d.spawnWindow * 100, 5, 100) / 100,
    );
  }

  static const GameRules defaults = GameRules();

  final double daySeconds;
  final double nightSeconds;

  /// zombies on night 1
  final int zombiesFirstNight;

  /// added every night after the first
  final int zombiesPerNight;

  /// seconds after sunset before the first group appears
  final double spawnDelaySeconds;

  /// the groups are spread over this share of the night, 0..1
  final double spawnWindow;

  double get cycleSeconds => daySeconds + nightSeconds;

  Map<String, Object?> toJson() => {
    'daySeconds': daySeconds,
    'nightSeconds': nightSeconds,
    'zombiesFirstNight': zombiesFirstNight,
    'zombiesPerNight': zombiesPerNight,
    'spawnDelaySeconds': spawnDelaySeconds,
    'spawnWindowPercent': (spawnWindow * 100).round(),
  };

  @override
  bool operator ==(Object other) =>
      other is GameRules &&
      other.daySeconds == daySeconds &&
      other.nightSeconds == nightSeconds &&
      other.zombiesFirstNight == zombiesFirstNight &&
      other.zombiesPerNight == zombiesPerNight &&
      other.spawnDelaySeconds == spawnDelaySeconds &&
      other.spawnWindow == spawnWindow;

  @override
  int get hashCode => Object.hash(
    daySeconds,
    nightSeconds,
    zombiesFirstNight,
    zombiesPerNight,
    spawnDelaySeconds,
    spawnWindow,
  );
}
